pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go

```diff
@@ -0,0 +1,156 @@
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+// serviceImportCache provides import state for all services in the cluster.
+type serviceImportCache interface {
+	GetClusterSetIPs(name types.NamespacedName) []string
+	HasSynced() bool
+	ImportedServices() []string
+}
+
+// newServiceImportCache creates a new cache of ServiceImport resources in the cluster.
+func newServiceImportCache(c *Controller) serviceImportCache {
+	if c.opts.EnableMCSServiceDiscovery {
+		informer := c.client.MCSApisInformer().Multicluster().V1alpha1().ServiceImports().Informer()
+		sic := &serviceImportCacheImpl{
+			Controller: c,
+			informer:   informer,
+			lister:     mcsLister.NewServiceImportLister(informer.GetIndexer()),
+		}
+
+		// Register callbacks for events.
+		c.registerHandlers(informer, "ServiceImports", sic.onEvent, nil)
+		return sic
+	}
+
+	// MCS Service discovery is disabled. Use a placeholder cache.
+	return disabledServiceImportCache{}
+}
+
+// serviceImportCacheImpl reads ServiceImport resources for a single cluster.
+type serviceImportCacheImpl struct {
+	*Controller
+	informer cache.SharedIndexInformer
+	lister   mcsLister.ServiceImportLister
+}
+
+func (ic *serviceImportCacheImpl) onEvent(obj interface{}, e model.Event) error {
+	si, ok := obj.(*mcsCore.ServiceImport)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return fmt.Errorf("couldn't get object from tombstone %#v", obj)
+		}
+		si, ok = tombstone.Obj.(*mcsCore.ServiceImport)
+		if !ok {
+			return fmt.Errorf("tombstone contained object that is not a ServiceImport %#v", obj)
+		}
+	}
+
+	// Update the cached service, if it exists.
+	ic.updateService(si, e)
+
+	// Trigger an XDS update.
+	ic.updateXDS(si)
+	return nil
+}
+
+func (ic *serviceImportCacheImpl) updateService(si *mcsCore.ServiceImport, e model.Event) {
+	// Extract the new IPs for the ClusterSet.
+	var ips []string
+	switch e {
+	case model.EventAdd, model.EventUpdate:
+		if si.Spec.Type == mcsCore.ClusterSetIP {
+			ips = si.Spec.IPs
+		}
+	}
+
+	// Update the cached service object, if it exists.
+	if svc, _ := ic.GetService(kubesr.ServiceHostnameForKR(si, ic.opts.DomainSuffix)); svc != nil {
+		svc.ClusterSetLocal.ClusterVIPs.SetAddressesFor(ic.Cluster(), ips)
+	}
+}
+
+func (ic *serviceImportCacheImpl) updateXDS(si *mcsCore.ServiceImport) {
+	hostname := kubesr.ServiceHostnameForKR(si, ic.opts.DomainSuffix)
+	pushReq := &model.PushRequest{
+		Full: true,
+		ConfigsUpdated: map[model.ConfigKey]struct{}{{
+			Kind:      gvk.ServiceEntry,
+			Name:      string(hostname),
+			Namespace: si.Namespace,
+		}: {}},
+		Reason: []model.TriggerReason{model.ServiceUpdate},
+	}
+	ic.opts.XDSUpdater.ConfigUpdate(pushReq)
+}
+
+func (ic *serviceImportCacheImpl) GetClusterSetIPs(name types.NamespacedName) []string {
+	si, err := ic.lister.ServiceImports(name.Namespace).Get(name.Name)
+	if err == nil && si.Spec.Type == mcsCore.ClusterSetIP && len(si.Spec.IPs) > 0 {
```

Comment:
IMO, you donot need so many conditions here, only  ClusterSetIP type can have ips set
if si !=nil {
 return si.Spec.IPs
}

Reply:
done.